### PR TITLE
Rust bindings

### PR DESCRIPTION
**Pull request by Daniel-Liu-c0deb0t**
## Description
Initial work on Rust bindings, based on issue #55.

~~For now, instead of Bazel, I am using Cargo to build the bindings, by manually linking the dependencies using `build.rs`. Unfortunately, it is not able to automatically link to the C and C++ `psi_server` and `psi_client` static libraries because they share the same name. To work around this, the direct path to the C++ version is directly passed as an argument to the linker. This can be ran with the `test.sh` bash script. Eventually, we need to figure out how to get Bazel to build the bindings. `build.rs` also generates the required protobuf code.~~

**Edit: Bazel can now be used to build the code: `bazel build //private_set_intersection/rust/...`** Linking to both the C and C++ versions of `psi_server` and `psi_client` still seemed to have naming conflicts, so the Bazel library names for the C bindings had to be changed to `c_psi_server` and `c_psi_client`.

This uses the C bindings instead of the C++ interface, which should be the recommended way for Rust FFI. `bindgen` is not currently used, because there aren't many functions, so hand-writing the bindings is feasible.

The API provided in the Rust bindings is slightly different than the APIs in the other languages, in order to better match the Rust style of doing things.

Stuff to do:
- [x] Protobufs
- [x] PSI server
- [x] PSI client
- [x] Bazel integration
- [x] Tests
- [x] Documentation

## Affected Dependencies
Added a Rust dependency on `libc=0.2.74` and `protobuf=2.8.2`. `cargo raze` was ran in the `third_party/cargo/` directory to generate Bazel build files for the Rust dependencies.

## How has this been tested?
Rust unit tests and integration tests were written, based off of the tests for the Go bindings.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -72,3 +72,19 @@ jobs:
         run: .github/workflows/scripts/run_tests_go.sh
       - name: Go linters
         run: .github/workflows/scripts/lint_go.sh
+
+  Rust:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [macos-latest, ubuntu-18.04]
```
Comment (madhavajay) [inline .github/workflows/tests.yml:80]: Should this be ubuntu-latest so that it gets updated down the track, or is this intentionally fixed to 18.04?

Comment (Daniel-Liu-c0deb0t) [reply .github/workflows/tests.yml]: Definitely wasn't intentional. I've updated it to use `ubuntu-latest`.

File: private_set_intersection/rust/src/client.rs [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+    pub fn create_request<T: AsRef<str>>(&self, raw_input: &[T]) -> ClientResult<Request> {
+        unsafe {
+            let input: Vec<PsiClientBuffer> = raw_input.iter().map(|s| PsiClientBuffer {
+                ptr: s.as_ref().as_ptr() as *const c_char,
+                len: s.as_ref().len() as size_t
+            }).collect();
+
+            let mut null_ptr: *mut c_char = ptr::null_mut();
+            let error_ptr = &mut null_ptr;
+            let mut null_ptr: *mut c_char = ptr::null_mut();
```
Comment (madhavajay) [inline private_set_intersection/rust/src/client.rs:105]: Is this an intentional shadow of the above? If they are separate for separate uses then maybe they should be named differently so this doesn't get accidentally changed.

Comment (Daniel-Liu-c0deb0t) [reply private_set_intersection/rust/src/client.rs:105]: These are separate instances of `null_ptr`. They serve no other purpose than being valid placeholder pointers that get replaced later on, so there isn't really any risk of them accidentally getting changed. I don't think a different name would be beneficial.

Comment (Daniel-Liu-c0deb0t) [reply private_set_intersection/rust/src/client.rs:105]: You are right, I will rename them.

**Comment by Daniel-Liu-c0deb0t** [comment]
I refactored the error handling code according to @madhavajay.

**Pull request merged**